clang/test/CodeGenSYCL/parallel_for_this_item.cpp

```diff
@@ -135,8 +152,10 @@ int main() {
     cgh.parallel_for<class RAT>(range<1>(1), [=](id<1> I) { f(); });
 
     // This kernel does not call sycl::this_item, but does call this_id
-    cgh.parallel_for<class FOX>(range<1>(1),
-                                [=](id<1> I) { this_id<1>(); });
+    cgh.parallel_for<class FOX>(range<1>(1), [=](id<1> I) { this_id<1>(); });
+
+    // This kernel calls sycl::this_item
+    cgh.parallel_for<class BEE>(range<1>(1), [=](auto I) { this_item<1>(); });
```

Comment:
Oh I see. So `getOperatorParens` doesn't return operator method when auto is used? Do you  understand why? It is not obvious to me why. @premanandrao could you also please take a look?

Reply:
@rdeodhar, is this the case that was failing before?

Reply:
Yes, this particular construct in an existing runtime test was failing. I have replicated the situation in the existing clang lit test. Owing to some templating of members, the operator() of the KernelObj was not being found. However, the code looking for the operator() is unnecessary because the kernel function is already available.

Reply:
Hmm... That means KernelObj is defined differently when parameter type is auto? Do you know why this happens? I'm approving the patch since the answer to this question is probably outside the scope of this patch. However, if you know the answer I am curious :)

Reply:
IMO, there is likely value to instead fix the poorly named getOperatorParens instead.

Reply:
FWIW, my first guess with getOperatorParens:

'auto' makes that a template-lambda.  This causes 2 CXXMethodDecls to exist in the type that are presumably both call operators.  I'm guessing you're getting the dependent one and that is messing with it.

Reply:
So I looked a bit, I was slightly wrong, the 2 CXXMethodDecls are actually wrapped in a FunctionTemplateDecl, so CXXRecordDecl::methods doesn't get them.  We'd probably have the same problem with a function-object with a templated operator() as well.

@AaronBallman : Can you suggest a way to fix https://github.com/intel/llvm/blame/sycl/clang/lib/Sema/SemaSYCL.cpp#L2079 to make sure we find the operator()?

I also want to know how we intend to actually USE that in the case where we have multiple operator()s defined in a type?  @kbobrovs, what happens then?

Reply:
Yep. I was just debugging this.  Dump of KernelObj - 

```
CXXRecordDecl 0x555564e958c8 <../parallel_for_this_item.cpp:158:46> col:46 implicit class definition
:
:
|-FunctionTemplateDecl 0x555564e95bb0 <<invalid sloc>, col:76> col:46 operator()
| |-TemplateTypeParmDecl 0x555564e956b8 <col:50, col:55> col:55 implicit class depth 0 index 0 I:auto
| |-CXXMethodDecl 0x555564e95b00 <col:56, col:76> col:46 operator() 'auto (auto) const' inline
| | |-ParmVarDecl 0x555564e95790 <col:50, col:55> col:55 I 'auto'
| | `-CompoundStmt 0x555564e96698 <col:58, col:76>
| |   `-CallExpr 0x555564e96678 <col:60, col:73> 'item<1>':'cl::sycl::item<1>'
| |     `-ImplicitCastExpr 0x555564e96660 <col:60, col:71> 'item<1> (*)()' <FunctionToPointerDecay>
| |       `-DeclRefExpr 0x555564e96608 <col:60, col:71> 'item<1> ()' lvalue Function 0x555564e8b688 'this_item' 'item<1> ()' (FunctionTemplate 0x555564e58ce8 'this_item')
| `-CXXMethodDecl 0x555564ea2d30 <col:56, col:76> col:46 used operator() 'void (cl::sycl::id<1>) const' inline
|   |-TemplateArgument type 'cl::sycl::id<1>'
|   | `-RecordType 0x555564e98b70 'cl::sycl::id<1>'
|   |   `-ClassTemplateSpecialization 0x555564e98a88 'id'
|   |-ParmVarDecl 0x555564ea2c18 <col:50, col:55> col:55 I 'cl::sycl::id<1>':'cl::sycl::id<1>'
|   `-CompoundStmt 0x555564ea2ec0 <col:58, col:76>
|     `-CallExpr 0x555564ea2ea0 <col:60, col:73> 'item<1>':'cl::sycl::item<1>'
|       `-ImplicitCastExpr 0x555564ea2e88 <col:60, col:71> 'item<1> (*)()' <FunctionToPointerDecay>
|         `-DeclRefExpr 0x555564ea2e30 <col:60, col:71> 'item<1> ()' lvalue Function 0x555564e8b688 'this_item' 'item<1> ()' (FunctionTemplate 0x555564e58ce8 'this_item')
`-CXXDestructorDecl 0x555564e966f0 <col:46> col:46 implicit referenced ~ 'void () noexcept' inline default trivial
```

Reply:
@erichkeane would you know why its getting wrapped?

Reply:
@erichkeane well that's a fun case -- I think you can use `decls_begin()` and `decls_end()` to do this:
```
static const CXXMethodDecl *getOperatorParens(const CXXRecordDecl *Rec) {
  auto MethIter = llvm::find_if(Rec->methods(), [](const CXXMethodDecl *MD) {
    return MD->getOverloadedOperator() == OO_Call;
  });
  if (MethIter != Rec->methods_end())
   return *MethIter;

  using function_template_iterator = specific_decl_iterator<FunctionTemplateDecl>;
  using function_template_range = llvm::iterator_range<specific_decl_iterator<FunctionTemplateDecl>>;
  auto FTIter = llvm::find_if(function_template_range(function_template_iterator(Rec->decls_begin()),
                                                      function_template_iterator(Rec->decls_end())),
                                           [](const FunctionTemplateDecl *FTD) {
                                             if (const auto *MD = dyn_cast<CXXMethodDecl>(FTD->getTemplatedDecl()))
                                               return MD->getOverloadedOperator() == OO_Call;
                                             return false;
                                           });
  if (FTIter != function_template_iterator(Rec->decls_end()))
   return *FTIter;
  return nullptr;
}
```
(This is totally untested and was written in a web browser, so YMMV.)

Reply:
> 
> 
> @erichkeane would you know why its getting wrapped?

Because we ALWAYS wrap templates.  We need something in the AST to contain BOTH the uninstantiated and instantiated template.

We should probably update the getOperatorParens, but, IMO, it should probably return an array of valid (non-dependent) operators.  I'm guessing this whole idea of using getOperatorParens is flawed though.  Trying to figure out the call operator of the type rather than checking the callgraph is likely the wrong way about it.  @kbobrovs : Please take a look at how you're using getOperatorParens and make sure it would work in a case where there is multiple operator()s, including instantiated template versions.

IN the case of what @AaronBallman just posted (thanks btw!), I don't know if that works for 2 reasons:  First, I think the `return *FTIter;` line is wrong, since that returns a `FunctionTemplateDecl*`, instead of a `CXXMethodDecl*`.  Second, it has the same problem of only returning the FIRST operator () :(

Reply:
Good call, that should be `return cast<CXXMethodDecl>(*FTIter)` instead, and it definitely only finds the first. As for how to handle multiple `operator()` definitions -- I agree that the function should either return a container of all the operators or take the container as a parameter.

Reply:
> @kbobrovs : Please take a look at how you're using getOperatorParens and make sure it would work in a case where there is multiple operator()s, including instantiated template versions.

AFAIR, this simplistic approach was based on assumption that the API code is under our control, and we know that there are no other operator '()' calls from the parallel_for_kernel.